### PR TITLE
Make `chainHead_unstable_storage` more powerful

### PR DESCRIPTION
**Pull request by tomaka**
Close https://github.com/paritytech/json-rpc-interface-spec/issues/21

Even though they're unrelated, this PR is based on top of https://github.com/paritytech/json-rpc-interface-spec/pull/35. I expect https://github.com/paritytech/json-rpc-interface-spec/pull/35 to be merged, and didn't want to have conflicts.

This PR enhances `chainHead_unstable_storage` by adding back the `type` field that was removed in #17.
You can now query not just the value of a storage item, but also the hash of its value, the values of the item and all its descendants, the hashes of item and all its descendants, or the Merkle value of the item or its closest ancestor.

Querying the descendants of the storage item makes it possible to enumerate all the keys under a certain prefix, which is important to know the content of maps.
In order to reproduce something similar to `state_getKeysPaged`, in the case when you enumerate descendants the server can now generate a `waiting-for-continue` event, after which the client must call `chainHead_unstable_storageContinue` in order to resume the enumeration. This makes it possible for the client to interrupt the enumeration at any point in time. Rather than letting the client decide of the number of items (like is the case for `state_getKeysPaged`), it's now the server which decides, which makes it possible to decide to generate this event based on the size of what was sent rather than the number of items.

The `closest-ancestor-merkle-value` query type makes it possible to know when the content of a map has changed. It can be seen as similar to `hash`, except that `hash` is only the hash of the value while `closest-ancestor-merkle-value` is the hash of the value plus the hash of all the descendants in the trie.
Basically, in order to watch a map, a JSON-RPC client would periodically query the `closest-ancestor-merkle-value` of the map, then redownload the list of keys if this Merkle value changes.

### CODE REVIEW

File: src/api/chainHead_unstable_storage.md [hunk 1]
```diff
@@ -5,19 +5,28 @@
...
+- `childTrie`: `null` for main storage look-ups, or a string containing the hexadecimal-encoded key of the child trie of the "default" namespace.
+- `type`: String equal to one of: `value`, `hash`, `closest-ancestor-merkle-value`, `descendants-values`, `descendants-hashes`.
 - `networkConfig` (optional): Object containing the configuration of the networking part of the function. See [here](./api.md) for details. Ignored if the JSON-RPC server doesn't need to perform a network request. Sensible defaults are used if not provided.
 
 **Return value**: String containing an opaque value representing the operation.
 
-The JSON-RPC server must start obtaining the value of the entry with the given `key` (and possibly `childKey`) from the storage.
+The JSON-RPC server must start obtaining the value of the entry with the given `key` from the storage, either from the main trie of from `childTrie`. If `type` is `descendants-value` or `descendants-hashes`, then it must also obtain the values of all the descendants of the entry.
```
Comment (lexnv) [inline src/api/chainHead_unstable_storage.md:14]:
```suggestion
The JSON-RPC server must start obtaining the value of the entry with the given `key` from the storage, either from the main trie or from `childTrie`. If `type` is `descendants-value` or `descendants-hashes`, then it must also obtain the values of all the descendants of the entry.
```

File: src/api/chainHead_unstable_storage.md [hunk 2]
```diff
@@ -35,18 +44,56 @@ This function will later generate notifications in the following format:
...
+If the `type` parameter was `"value"`, `"hash"`, `"descendants-values"` or `"descendants-hashes"`, this `key` is guaranteed to start with the `key` provided as parameter.
+If the `type` parameter was `"value"` or `"hash"`, then it is also guaranteed to be equal to the `key` provided as parameter.
+If the `type` parameter was `"closest-ancestor-merkle-value"`, then the `key` provided as parameter is guaranteed to start with the value in the `key` field.
+
+If the `type` parameter was `"value"` or `"descendants-values"`, then the `value` field is set. The `value` field a string containing the hexadecimal-encoded value of the storage entry.
+
+If the `type` parameter was `"hash"` or `"descendants-hashes"`, then the `hash` field is set. The `hash` field a string containing the hexadecimal-encoded hash of the storage entry.
+
+If the `type` parameter was `"closest-ancestor-merkle-value"`, then the `merkle-value` field is set and the `key` field indicates which closest ancestor has been found. The `merkle-value` field a string containing the hexadecimal-encoded Merkle value of the storage item indicated by the `key` field.
```
Comment (lexnv) [inline src/api/chainHead_unstable_storage.md:70]: small nit:
```suggestion
If the `type` parameter was `"value"` or `"descendants-values"`, then the `value` field is set. The `value` field is a string containing the hexadecimal-encoded value of the storage entry.

If the `type` parameter was `"hash"` or `"descendants-hashes"`, then the `hash` field is set. The `hash` field is a string containing the hexadecimal-encoded hash of the storage entry.

If the `type` parameter was `"closest-ancestor-merkle-value"`, then the `merkle-value` field is set and the `key` field indicates which closest ancestor has been found. The `merkle-value` field is a string containing the hexadecimal-encoded Merkle value of the storage item indicated by the `key` field.
```

**Review by lexnv — APPROVED** [review]
Amazing! 

This should give us parity between what substrate exposed previously wrt iterating over key storages, while having a more compact and robust API.

Hopefully, we'll be able to make a complete transition in the whole ecosystem to the new methods and deprecate the old RPC ones 👍

**Comment by tomaka** [comment]
I'm actually not completely happy with this design where the client has to call `storageContinue`.

The function achieves all its goals, but the problem is that the client can delay calling `storageContinue` indefinitely. There exists a limit to the number of active subscriptions, so I'm not worried about DoS attacks or anything like that.

What I'm worried about it a very specific situation: if a load balancer redirects a `storage` query to server A, and later wants to shut down server A because it is no longer needed or needs maintenance, the load balancer currently can generate a `stop` event for all active `follow` subscriptions, wait for all active requests to finish, then shut down server A.

But with the introduction of this `storageContinue` thing, the client can prevent this from happening by keeping a `storage` query active indefinitely.
I feel like there should be something similar to `stop` that the load balancer can generate. The `error` event currently means "fatal irrecoverable error", which isn't the case here as the client should simply try again. I think that `inaccessible` is appropriate, because the load balancer can now longer access the node in question.

**Comment by tomaka** [comment]
I just wanted to mention that I can't merge this pull request myself as I'm not at Parity anymore.

**Review by lexnv — APPROVED** [review]
LGTM!

// cc @skunert @jsdw @bkchr

**Review by skunert — APPROVED** [review]

File: src/api/chainHead_unstable_storage.md [hunk 3]
```diff
@@ -5,19 +5,28 @@
...
+
+If `type` is `hash` or `descendants-hashes`, then the cryptographic hash of each item is provided rather than the full value. The hashing algorithm used is the one of the chain, which is typically blake2b. This can lead to significantly less bandwidth usage and can be used in order to compare the value of an item with a known hash and querying the full value only if it differs.
+
+If `type` is `closest-ancestor-merkle-value`, then the so-called trie Merkle value of the `key` is provided. If `key` doesn't exist in the trie, then the Merkle value of the closest ancestor of `key` is provided. Contrary to `hash`, a `closest-ancestor-merkle-value` always exists for every `key`. The Merkle value is similar to a hash of the value and all of its descendants together.
```
Comment (bkchr) [inline src/api/chainHead_unstable_storage.md:26]: If I understand it correctly, I find the naming "hard/confusing". Why not "closest-ancestor-merkle-node" and then also saying that this is the hash of the closest merkle node.

Comment (bkchr) [reply src/api/chainHead_unstable_storage.md:26]: > I kept the same vocabulary as the spec.

Okay :+1: 

> (the merkle value is either the node value or its hash).

What is the node value? The value attached to a node? Shouldn't we always be interested in the `hash` as we only use this for tracking changes?

Comment (tomaka) [reply src/api/chainHead_unstable_storage.md:26]: If the node value is less than 32 bytes than the Merkle value is the same as the node value. If the node value is 32 bytes or more, then it's the hash.
https://spec.polkadot.network/chap-state#defn-merkle-value

Comment (bkchr) [reply src/api/chainHead_unstable_storage.md:26]: Ahh okay, ty for the explanation :+1:
